### PR TITLE
To add a Garbage collection

### PR DESCRIPTION
**Pull request by sdminonne**
@clamoriniere1A @EricMountain-1A a mind validate the approach?

### CODE REVIEW

File: pkg/garbagecollector/garbagecollector.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+	jobs, err := c.KubeClient.BatchV1().Jobs(metav1.NamespaceAll).List(metav1.ListOptions{
+		LabelSelector: controller.WorkflowLabelKey,
+	})
+	if err != nil {
+		glog.Errorf("Garbage collector: unable to list workflow jobs to be collected: %v", err)
+	}
+	collected := make(map[string]struct{})
+	for _, job := range jobs.Items {
+		workflowName, found := job.Labels[controller.WorkflowLabelKey]
+		if !found {
```
Comment (dbenque) [inline pkg/garbagecollector/garbagecollector.go]: Does this test make sense? if it fails it means that the Jobs lister did return an items that does not respect the LabelSelector... any corner case that I am missing here?

File: pkg/garbagecollector/garbagecollector.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+			glog.Errorf("unable to find workflow name for job: %s/%s", job.Namespace, job.Name)
+			continue
+		}
+		if _, done := collected[path.Join(job.Namespace, workflowName)]; done {
+			continue // already collected so skip
+		}
+		if _, err := c.WorkflowLister.Workflows(job.Namespace).Get(workflowName); err == nil || !apierrors.IsNotFound(err) {
+			if err != nil {
+				glog.Errorf("unable to retrieve workflow %s/%s cache: %v", job.Namespace, workflowName, err)
```
Comment (dbenque) [inline pkg/garbagecollector/garbagecollector.go]: A comment like "Unexpected error while using the workflow cache for %s/%s: %v"
Because the IsNotFound error would also be a valid error for your comment...

To make things event more readable I would code it that way:
_, err := c.WorkflowLister.Workflows(job.Namespace).Get(workflowName);
if err== nil {
continue // the workflow is still valid
}
if !apierrors.IsNotFound(err) {
glog.Errorf("Unexpected error while using the workflow cache for %s/%s: %v")
continue
}


Even better maybe with a switch on err, but it will start to be a boomrang code.

File: pkg/garbagecollector/garbagecollector.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+			}
+			continue
+		}
+		// Workflow couldn't be find in cache. Tyring to get it via APIs.
+		if _, err := c.WorkflowClient.Workflow().Workflows(job.Namespace).Get(workflowName, metav1.GetOptions{}); err != nil {
+			if apierrors.IsNotFound(err) {
+				// then remove all the jobs.
+				if err := c.KubeClient.Batch().Jobs(job.Namespace).DeleteCollection(controller.CascadeDeleteOptions(0), metav1.ListOptions{
+					LabelSelector: controller.WorkflowLabelKey + "=" + workflowName}); err != nil {
+					if err != nil {
```
Comment (dbenque) [inline pkg/garbagecollector/garbagecollector.go]: Why doing that err!=nil test again here ? it must be true

File: pkg/garbagecollector/garbagecollector.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+						glog.Errorf("unable to delete Collection of jobs for workflow %s/%s", job.Namespace, workflowName)
+					}
+				}
+				collected[path.Join(job.Namespace, workflowName)] = struct{}{}
+				glog.Infof("Removed all jobs for workflow %s/%s", job.Namespace, workflowName)
+				continue
```
Comment (dbenque) [inline pkg/garbagecollector/garbagecollector.go]: useless continue statement

File: pkg/garbagecollector/garbagecollector.go (hunk 1 above)
Comment (sdminonne) [reply pkg/garbagecollector/garbagecollector.go]: The day someone will break Generalized Label Selector for CRD it will break.

File: pkg/garbagecollector/garbagecollector.go (hunk 3 above)
Comment (sdminonne) [reply pkg/garbagecollector/garbagecollector.go]: right thanks

File: pkg/garbagecollector/garbagecollector.go (hunk 2 above)
Comment (sdminonne) [reply pkg/garbagecollector/garbagecollector.go]: The idea here is that we look for jobs without a workflow. So we're interested in case where there's error AND the error IsNotFound for DeMorgan law it's what we test in-line... Finally it's the same :)

But it's generally true for the comment. Going to change.  Thanks

**Comment by sdminonne** [comment]
Folks following the LGTM from @clamoriniere1A  I would merge it. Let me know if its' good for you

**Comment by sdminonne** [comment]
@alexei-led as you asked this implements GC

File: test/e2e/workflows.go [hunk 5]
```diff
@@ -36,13 +36,12 @@ func deleteAllJobs(kubeClient clientset.Interface, workflow *wapi.Workflow) {
 		return
 	}
 	for i := range jobs.Items {
-		kubeClient.Batch().Jobs(workflow.Namespace).Delete(jobs.Items[i].Name, cascadeDeleteOptions(0))
+		kubeClient.Batch().Jobs(workflow.Namespace).Delete(jobs.Items[i].Name /*cascadeDeleteOptions(0)*/, nil)
```
Comment (clamoriniere1A) [inline test/e2e/workflows.go]: this dead code can be removed

Comment (sdminonne) [reply test/e2e/workflows.go]: ouch.... thanks. In a follow-up!
